ruma-events/src/pdu.rs

```diff
@@ -0,0 +1,934 @@
+//! Types for persistent data unit schemas
+//!
+//! The differences between the `RoomV1Pdu` schema and the `RoomV3Pdu` schema are
+//! that the `RoomV1Pdu` takes an `event_id` field (`RoomV3Pdu` does not), and
+//! `auth_events` and `prev_events` take `Vec<(EventId, EventHash)> rather than
+//! `Vec<EventId>` in `RoomV3Pdu`.
+//!
+//! The stubbed versions of each PDU type remove the `event_id` field (if any)
+//! and the `room_id` field for use in PDU templates.
+
+use std::{collections::BTreeMap, time::SystemTime};
+
+use js_int::UInt;
+use ruma_events::EventType;
+use ruma_identifiers::{EventId, RoomId, UserId};
+use serde::{Deserialize, Serialize};
+use serde_json::Value as JsonValue;
+
+/// Enum for PDU schemas
+#[derive(Clone, Debug, Deserialize, Serialize)]
+#[serde(untagged)]
+pub enum Pdu {
+    /// PDU for room versions 1 and 2.
+    RoomV1Pdu(RoomV1Pdu),
+    /// PDU for room versions 3 and above.
+    RoomV3Pdu(RoomV3Pdu),
+}
+
+/// A 'persistent data unit' (event) for room versions 1 and 2.
+#[derive(Clone, Debug, Deserialize, Serialize)]
+pub struct RoomV1Pdu {
+    /// Event ID for the PDU.
+    pub event_id: EventId,
+
+    /// The room this event belongs to.
+    pub room_id: RoomId,
+
+    /// The user id of the user who sent this event.
+    pub sender: UserId,
+
+    /// The `server_name` of the homeserver that created this event.
+    pub origin: String,
+
+    /// Timestamp (milliseconds since the UNIX epoch) on originating homeserver
+    /// of when this event was created.
+    #[serde(with = "ruma_serde::time::ms_since_unix_epoch")]
+    pub origin_server_ts: SystemTime,
+
+    // TODO: Replace with event content collection from ruma-events once that exists
```

Comment:
As discussed elsewhere, `Pdu`s should probably be defined similarily to [`AnyRoomEvent`](https://github.com/ruma/ruma/blob/9074e4c17a3751ca740916823863f07ca8334f7d/ruma-events/src/enums.rs#L139) eventually. Can you update this comment (and the other ones in this module) accordingly?

Reply:
I've been pretty out of the loop with the event types; what do you want me to put there exactly? That EventType should be changed to AnyRoomEvent once it's merged?

Reply:
The pdu structs will probably be split further into sub-types (you really want a redaction pdu to always have `redacts`, not optionally) and will no longer have an explicit `event_type` field, the variant will instead be encoded in enum variants.

Maybe

```rust
// TODO: Encode event type as content enum variant, like event enums do
```

would be good.